Sources/SwiftDocC/Semantics/MarkupReferenceResolver.swift

```diff
@@ -113,9 +113,13 @@ struct MarkupReferenceResolver: MarkupRewriter {
         guard let resolvedURL = resolve(reference: unresolved, range: link.range, severity: .warning) else {
             return link
         }
-        var link = link
-        link.destination = resolvedURL.absoluteString
-        return link
+        if link.destination != resolvedURL.absoluteString {
+            var link = link
+            link.destination = resolvedURL.absoluteString
+            return link
+        } else {
+            return link
```

Comment:
Not sure I understand why this change is needed, can you elaborate?

Reply:
Only update the destination when needed. Because when we update link.destination we'll lose the parsedRange.

+1 for this. This seems not be needed anymore for me since we will preserve the parsedRange in https://github.com/apple/swift-markdown/pull/82 or we will find other ways to different `<doc://xx>` and `[]()`

Reply:
> Only update the destination when needed. Because when we update link.destination we'll lose the parsedRange.
> 
> +1 for this. This seems not be needed anymore for me since we will preserve the parsedRange in [apple/swift-markdown#82](https://github.com/apple/swift-markdown/pull/82) or we will find other ways to different `<doc://xx>` and `[]()`

After reviewing the [CMark Spec](https://github.com/github/cmark-gfm/blob/master/test/spec.txt), I discovered that it does not have any specially treatment for `<doc://xx>` in markdown. Therefore, no modifications are necessary on the markdown side. It's worth noting that the `<xx>` link is simply a syntax for creating links and functions identically to `[xx](xx)`.
```
## Autolinks

[Autolink](@)s are absolute URIs and email addresses inside
`<` and `>`. They are parsed as links, with the URL or email address
as the link label.
```

Reply:
The issue lies with `swift-docc` side. Using `overridingTitle` directly will result in the same incorrect output for both formats.
```
- [doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))
- <doc:Sloth/eat(_:quantity:)>
```

For the above

Reply:
> The issue lies with `swift-docc` side. Using `overridingTitle` directly will result in the same incorrect output for both formats.
> 
> ```
> - [doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))
> - <doc:Sloth/eat(_:quantity:)>
> ```

Before the PR, for both formats, we'll get `overridingTitle: nil` & `overridingTitleInlineContent: nil` and encode it to the following json for swift-docc-render to consume. 
```
"content": [
    {
        "type": "paragraph",
        "inlineContent": [
            {
                "type": "reference",
                "isActive": true,
                "identifier": "doc://SlothCreator/documentation/SlothCreator/Sloth/eat(_:quantity:)"
            }
        ]
    }
]
```
Then swift-docc-render will use the reference's titleInlineContent
https://github.com/apple/swift-docc-render/blob/eed919f5dcd8993455dea6534f3c051d58e1cfb3/src/components/ContentNode.vue#L429-L430

If we'd like to support [custom title](doc:xx). A breaking change is needed here since `[doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))` and `<doc:Sloth/eat(_:quantity:)>` is basically identical outside the CMark Parser world.(Unless we hardcode to check the first char is `[` or `<`) 

So after the PR
`<doc:Sloth/eat(_:quantity:)>` will produce the following node json content
```
"content": [
    {
        "type": "paragraph",
        "inlineContent": [
            {
                "type": "reference",
                "isActive": true,
                "identifier": "doc://SlothCreator/documentation/SlothCreator/Sloth/eat(_:quantity:)",
                "overridingTitle": "doc:Sloth/eat(_:quantity:)"
            }
        ]
    }
]
```
and then swift-docc-render will use "doc:Sloth/eat(_:quantity:)" as its title instead of "eat(_:quantity:)".

Reply:
Hi @franklinsch, I am seeking your assistance with a design decision for the PR. Could you kindly review it and provide your feedback on the proposed changes? Thank you in advance for your help.

1. Both formats will give the same standard overriding output of "`doc:Sloth/eat(_:quantity:)`"
- Pros: Fixing the custom doc link title issue
- Cons: Making `<doc:Sloth/eat(_:quantity:)>` show "`doc:Sloth/eat(_:quantity:)`" instead of "`eat(_:quantity:)`"
```
- [doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))
- <doc:Sloth/eat(_:quantity:)>
```

2. Do a specially treatment for <doc://>
- Pros: Fixing the custom doc link title issue while maintaining the history behavior for `<doc:Sloth/eat(_:quantity:)>`
- Cons: We are adding strange logic for the CMark Spec and add maintain cost.
  - According to the CMark spec `<doc:Sloth/eat(_:quantity:)>` is just a syntax sugar for `[doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))`

Reply:
thanks for the overview. I would much prefer if we implemented a solution that doesn't break existing behaviour. It seems like Suggestion A would change the way links get rendered.

For suggestion B, is there a way we can isolate the special behaviour to just auto links using the `doc` URL scheme? I'm fine with `<doc:Sloth/eat(_:quantity:)>` getting translated to `[eat(_:quantity:)](doc:Sloth/eat(_:quantity:)`. Turning `doc` links into links whose title is the page's title is one of DocC's features, so this transformation makes sense. You mention adding special logic for the CMark spec; where would these changes be implemented? Swift-Markdown and Swift-DocC?

Reply:
Swift-DocC side. And the implementation will look similar to [this](https://github.com/apple/swift-markdown/blob/4cbd46eab22cbeef307095d17958b6ed23b5f38f/Sources/Markdown/Walker/Walkers/MarkupFormatter.swift#L819-L849)

We check the **doc host + autolinks** format - `<doc://>` and assign overridingTitle to nil to match the original behaviour

Reply:
I was thinking to add a style property to `Markdown.Link`. But eventually failed to do so since the underlying `cmark-gfm` parser does not give us the extra info. There is only one node type - CMARK_NODE_LINK
```
public struct Link: InlineMarkup, InlineContainer {
    /// The style of the link
    public var style: Style = .normal
    
    public enum Style {
        /// Normal reference link
        ///
        /// For example `[example text](http://example.com)`
        case normal
        /// Doc style link
        ///
        /// For example `<doc://example>`
        case doc
    }
    ...
}
```

> Another implementation we can reference is swift-cmark's [is_autolink](https://github.com/apple/swift-cmark/blob/86aeb491675de6f077a3a6df6cbcac1a25dcbee1/src/commonmark.c#L152-L154)

Reply:
Ah I see. If checking if the text of the link matches its destination is a common way of indicating that a link is an auto-link, it makes sense to adopt the same behaviour then. We could expose a computed property on `Markdown.Link`.

Reply:
> Gotcha. The check we're doing in Markdown to determine whether a `Link` is an auto link or not seems finicky (`destination == text.string`). If there's a way we can add a property to `Link` that indicates what style of link it is that we can populate during parsing, that would be more robust.

Yes, this check is finicky since it will treat `[doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))` as an auto-link too. Not respect the user's overriding title "`doc:Sloth/eat(_:quantity:)`".

If we really do not want to break existing behavior. The only solution I came out to diff them is use parsed range to compare "[" and "]" char.
Or the end user of DocC will lose the ability to write a link whose destination is `doc:Sloth/eat(_:quantity:)` and title is `doc:Sloth/eat(_:quantity:)`.(Using the Markdown implementation)

```swift
let source = #"""
<doc://example.com>[doc://example.com](doc://example.com)
"""#
let document = Document(parsing: source, options: [])
print(document.debugDescription())
```
```
Document
└─ Paragraph
   ├─ Link destination: "doc://example.com"
   │  └─ Text "doc://example.com"
   └─ Link destination: "doc://example.com"
      └─ Text "doc://example.com"
```

Reply:
> Ah I see. If checking if the text of the link matches its destination is a common way of indicating that a link is an auto-link, it makes sense to adopt the same behaviour then. We could expose a computed property on `Markdown.Link`.

Got it. I'll update to use this approach. The only risk left is the end user of DocC will lose the ability to write a link whose destination is `doc:Sloth/eat(_:quantity:)` and title is `doc:Sloth/eat(_:quantity:)`. 

Since `[doc:Sloth/eat(_:quantity:)](doc:Sloth/eat(_:quantity:))` will be treated as an auto link and the host is `doc`, we'll make the overridingTitle to nil.

Reply:
> Got it. I'll update to use this approach. The only risk left is the end user of DocC will lose the ability to write a link whose destination is doc:Sloth/eat(_:quantity:) and title is doc:Sloth/eat(_:quantity:).

I'm not super concerned about this; I don't really see why a user would want to display `doc:Sloth/eat(_:quantity:)` in plain text. In the worst case, they can embed the text in backticks.